subsys/net/l2/ethernet/ethernet.c

```diff
@@ -518,7 +518,8 @@ static struct net_buf *ethernet_fill_header(struct ethernet_context *ctx,
 	}
 
 	if (IS_ENABLED(CONFIG_NET_VLAN) &&
-	    net_eth_is_vlan_enabled(ctx, net_pkt_iface(pkt))) {
+	    net_eth_is_vlan_enabled(ctx, net_pkt_iface(pkt)) &&
+	    ptype != htons(NET_ETH_PTYPE_PTP)) {
```

Comment:
1. It's not so nice to introduce a reference to a higher protocol (PTP) here as it breaks encapsulation / proper OSI layering. I know that we've plenty of precedences for this in other places in the net stack, but we should try removing those instances rather than introducing more. We should find a proper mechanism to invert the layering IMO (i.e. let the higher protocol control VLAN header insertion rather than the lower protocol having to check for presence of any arbitrary higher layer protocol). @jukkar If you agree, do you have any idea how we could invert control here?

2. Does it make sense to also add && !IS_ENABLED(CONFIG_NET_GPTP_VLAN) such that the (non-standard) PTP over VLAN option is not being compromised? This would still remove the header if that option is off but leave it there if it's on.

Reply:
Even though the VLAN tagging is not used for gPTP, it can be useful in testing purposes (to isolate the gPTP traffic), so I would like to keep it for that. Perhaps that could be described in the Kconfig better (it is not mentioned at all that VLAN tagging is not used in gPTP).

@fgrandel yes, that is a good point about layering. As for the gPTP traffic, it is L2 protocol with its own Ethernet protocol type so it kind of makes sense to have in ethernet.c

Reply:
From my side: yes.

Reply:
done